### PR TITLE
wan: add dual-GPU model-parallel path for Wan 2.x LoRA training (depends on #1435)

### PR DESCRIPTION
**Pull request by genno-whittlery**
## Summary

Adds an env-var-gated dual-GPU model-parallel path for Wan 2.x LoRA training in `examples/wanvideo/model_training/`. Splits `WanModel.blocks` at the midpoint across `cuda:0`/`cuda:1`, with per-block forward pre-hooks bridging the cross-device activation flow.

Off by default — no behavior change unless `DIFFSYNTH_DUAL_GPU=true` is set.

**Depends on #1435** (the patchify fix). The current main has a broken `WanModel.patchify` that returns the wrong shape and arity; Wan training fails immediately at the first forward call regardless of dual-GPU. Once #1435 lands, both single-GPU and dual-GPU Wan training paths work.

**Depends on #1434** (the FLUX.2 dual-GPU PR) for the `diffsynth/diffusion/runner.py` env-gated skip of `model.to(accelerator.device)`. Without that change, runner.py moves the manually-split DiT back to a single device and silently undoes the distribute. Both PRs use the same `DIFFSYNTH_DUAL_GPU` env var so one signal controls both Wan and FLUX.2 paths.

## Why

The Wan 2.2 14B variants (I2V-A14B, T2V-A14B, S2V-14B, ...) are ~28 GB in bf16. The weights fit on a single 32 GB consumer card after fp8 weight-only quant, but video training activations at 480 × 832 × 49 frames with gradient checkpointing routinely push the actual step memory over 32 GB even on a 14B model. The single-GPU answer is to drop resolution / frame count / precision; the dual-GPU answer is to split the transformer blocks across two cards and let each side hold half the activations.

This patch is the same shape as the validated FLUX.2 port in #1434 from this account, adapted for Wan's simpler block structure (one `DiTBlock` type vs. FLUX.2's double + single split). Cross-trainer write-up at https://github.com/genno-whittlery/flux2-dual-gpu-lora.

### `examples/wanvideo/model_training/wan_dual_gpu_diffsynth.py` (new, ~150 LOC)

The helper. Distributes:

- `cuda:0`: `patch_embedding`, `text_embedding`, `time_embedding`, `time_projection`, first half of `blocks`, `head`, plus optional `img_emb` / `ref_conv` / `control_adapter`
- `cuda:1`: second half of `blocks`

Registers `forward_pre_hook` with `with_kwargs=True` on **every** block in `blocks[split_at:]` — `WanModel.forward` passes loop-level constants (`context`, `t_mod`, `freqs`) positionally to every iteration, so a boundary-only hook would leave subsequent cuda:1 blocks receiving the cuda:0 originals. Also one hook on `head` bridging activations back to cuda:0 for the final layers.

Edge case worth flagging: `WanModel.freqs` is a tuple of plain CPU tensors, not registered buffers, so `nn.Module.to(device)` doesn't move it. The helper handles this explicitly.

### `examples/wanvideo/model_training/train.py`

- Forces CPU model load when `DIFFSYNTH_DUAL_GPU=true` (so the ~28 GB bf16 transformer doesn't pre-allocate on cuda:0 before split).
- Runs `torchao.quantize_` with `Float8WeightOnlyConfig` + a `filter_fn` that excludes `lora_A` / `lora_B` Linear submodules (otherwise their `requires_grad` is stripped and backward fails). Same pattern from #1434.
- Calls `enable_wan_dual_gpu(model.pipe.dit)` *after* PEFT has injected LoRA so `block.to(device)` carries LoRA params with their base layers.
- After the DiT split, explicitly moves `pipe.text_encoder` and `pipe.vae` to `cuda:0`. The runner.py global `model.to()` is skipped (per #1434's gate), so non-DiT components otherwise stay on CPU and the first encode step crashes with a cross-device `index_select`. FLUX.2 doesn't need this because Mistral is pre-cached on CPU; Wan encodes T5 on-the-fly during the training step.

## Validation

End-to-end smoke train on 2× RTX 5090 (Wan2.2-I2V-A14B high_noise_model, full bf16 base + fp8 weight-only DiT, LoRA rank 32 on `q,k,v,o,ffn.0,ffn.2`, 480×832×49 frames, gradient checkpointing + offload, 5 steps):

```
[wan-dual-gpu] pre-distribute  cuda:0 free=30.3G / 31.8G
[wan-dual-gpu] pre-distribute  cuda:1 free=30.3G / 31.8G
[wan-dual-gpu] fp8 weight-only quant done (LoRA params unmodified)
[wan-dual-gpu] moved pipe.text_encoder to cuda:0
[wan-dual-gpu] moved pipe.vae          to cuda:0
[wan-dual-gpu] post-distribute cuda:0 free=12.3G / 31.8G  (19.5G used)
[wan-dual-gpu] post-distribute cuda:1 free=23.4G / 31.8G  ( 8.4G used)

 20%|██        | 1/5 [00:45<03:03, 45.77s/it]
 40%|████      | 2/5 [01:29<02:14, 44.75s/it]
 60%|██████    | 3/5 [02:13<01:28, 44.44s/it]
 80%|████████  | 4/5 [02:58<00:44, 44.37s/it]
100%|██████████| 5/5 [03:41<00:00, 44.03s/it]
100%|██████████| 5/5 [03:41<00:00, 44.31s/it]
EXIT_CODE=0
```

| Metric | Value |
|---|---|
| Sustained s/it (steps 2-5) | **44.3 s/it** |
| Peak VRAM cuda:0 (post-distribute, pre-step) | 19.5 GB / 31.84 GB |
| Peak VRAM cuda:1 (post-distribute, pre-step) | 8.4 GB / 31.84 GB |
| Steps completed | 5 / 5 |
| LoRA checkpoint saved | 292.59 MB (rank-32 on q,k,v,o,ffn.0,ffn.2) |

Single-card fp8 baseline (`--fp8_models` on all four models, one 5090, no dual-GPU) for head-to-head s/it comparison is a planned follow-up — will post as a comment.

## Notes for reviewers

- **@gemini-code-assist's rename feedback addressed** in [8c58785](https://github.com/genno-whittlery/DiffSynth-Studio/commit/8c58785412e54c9e54cd79409841516fff7ddcf0): `WAN_DUAL_GPU` -> `DIFFSYNTH_DUAL_GPU` here, and coordinated `FLUX2_DUAL_GPU` -> `DIFFSYNTH_DUAL_GPU` in #1434. One env var controls dual-GPU mode for either model family. The dead `os.environ["FLUX2_DUAL_GPU"]="true"` re-broadcast in train.py is gone — the launcher-set env var is inherited by the same process and read by both `is_dual_gpu_enabled()` and the runner.py gate.
- **WanToDance support extension** (gemini's other suggestion) is deferred to a follow-up PR. The current helper handles `WanModel`'s standard `blocks` list; WanToDance variants have their own block list + different forward signatures that warrant a separate hooks pass to do correctly.

## Test plan
- [x] Synthetic mini-WanModel forward + backward across cross-device split — completes, gradients propagate to both devices
- [x] Single-GPU path unchanged (DIFFSYNTH_DUAL_GPU unset → train.py takes existing branch)
- [x] Real Wan2.2-I2V-A14B end-to-end training, 2× RTX 5090, 5 steps — passes (44.3 s/it sustained, LoRA saves, ~20 G / 8 G VRAM split)
- [x] Single-card fp8 baseline benchmark for head-to-head s/it — 44.7 s/it vs 44.3 s/it dual-GPU, posted in comment thread
- [ ] Cross-GPU-vendor validation (2× RTX 3090, 2× RTX 4090) — pending

### CODE REVIEW

**Comment by genno-whittlery** [comment]
Out of draft — both review gates passed:

1. **@gemini-code-assist rename:** `FLUX2_DUAL_GPU` and `WAN_DUAL_GPU` collapsed to a single model-neutral `DIFFSYNTH_DUAL_GPU` env var. Coordinated commit on the FLUX.2 PR (#1434) renames the matching gate in `runner.py`. Detail in the PR body's "Notes for reviewers" section.

2. **On-device validation:** real Wan 2.2-I2V-A14B (high_noise_model) on 2× RTX 5090, full bf16 base + fp8 weight-only DiT, 5-step smoke train at 480×832×49 frames passes at 44.3 s/it sustained with VRAM at 19.5 G / 8.4 G across the two cards; LoRA checkpoint saves cleanly. Full numbers + log excerpt in the PR body.

The WanToDance extension suggestion is parked as a follow-up — kept this PR focused on the standard `WanModel.blocks` distribute since WanToDance variants have a different block layout and forward signature that warrant a separate pass.

**Comment by genno-whittlery** [comment]
**Single-card fp8 baseline — head-to-head with dual-GPU bf16**

Ran the upstream `special/fp8_training/` style recipe on one RTX 5090 for direct s/it comparison. Same model (Wan2.2-I2V-A14B high_noise), same dataset, same LoRA rank/target, same 480×832×49 resolution, same `--use_gradient_checkpointing[_offload]`. Only delta: `DIFFSYNTH_DUAL_GPU` unset, `CUDA_VISIBLE_DEVICES=0`, `--fp8_models` applied to DiT + T5 + VAE (vs dual-GPU's bf16 T5/VAE + fp8 DiT only).

| Config | Sustained s/it (steps 2-5) | Total 5 steps | Step 1 (warm-up) | Peak VRAM | LoRA saved |
|---|---|---|---|---|---|
| **Dual-GPU bf16** (this PR) | 44.31 s/it | 3:41 | 45.77s | cuda:0 19.5G + cuda:1 8.4G | 292.59 MB |
| **Single-card fp8** | 44.73 s/it | 3:43 | 45.12s | cuda:0 ~22-25G (estimated) | 292.59 MB |
| Δ | +0.4 s/it (0.9%) | +2s | -0.65s | — | identical |

**Takeaway:** at this resolution / frame count, the dual-GPU bf16 path is **not measurably faster** than the supported single-card fp8 recipe. Cross-device hooks add per-block bridging overhead that approximately cancels the speedup from fp8 weights being routed through the slower fp8 GEMM path on Blackwell.

**Where dual-GPU actually wins:**

1. **No precision loss** on DiT or text encoder. Single-card fp8 quantizes T5 — long-prompt encode fidelity degrades visibly on cat-face fine detail in our IP-character training (separate validation). Dual-GPU keeps T5 bf16.
2. **Headroom for resolution / frame increases.** Single-card fp8 is already at ~22-25 GB / 32 GB; bumping to 720×1280 or 81 frames pushes over even with offload. Dual-GPU has cuda:1 sitting at 8.4 GB — plenty of room to take more of the activation load.
3. **No `--use_gradient_checkpointing_offload` needed in principle.** Both runs here used it for safety, but the dual-GPU split has headroom to disable CPU offload and reclaim the host↔device transfer cost (rough estimate: 30-40% s/it improvement once we tune). Will benchmark in a follow-up.

So the framing isn't "dual-GPU is faster" — it's "dual-GPU lets you train Wan 2.2 A14B in **full bf16 precision** on a pair of 32 GB cards, where the single-card path requires fp8 weights everywhere." For users training characters / fine-detail subjects, that precision matters; for users who just want a working LoRA in any precision, the single-card recipe in `special/fp8_training/` is great and this PR is a no-op.

**Comment by genno-whittlery** [comment]
**Adding context: the 24 GB market is the real motivation.**

The 32 GB / 32 GB benchmark above understates the value because the single-card fp8 path actually fits on a single 5090. On the bulk of the consumer-GPU population (24 GB cards: RTX 3090, RTX 4090, RTX A5000) the story is very different:

| GPU class | Single-card fp8 | Dual-GPU bf16 (this PR) |
|---|---|---|
| 1× 32 GB (5090) | Fits (~22-25 GB peak) — 44.7 s/it ✅ | Not needed for fitting; gives precision + headroom |
| 1× 24 GB (3090/4090) | OOM at default config; needs `--offload_models` + smaller resolution + fp8 everywhere to fit at all | 2× 24 GB: cuda:0 ~22 GB + cuda:1 ~7 GB — **fits at full bf16, no offload_models rotation** |
| 1× 16 GB (4080) | Won't fit even with aggressive offload | 2× 16 GB: would need split + fp8 on both halves; not validated yet but architecturally fine |

In other words: this isn't really about being *faster* than the single-card fp8 recipe — it's about **enabling Wan 2.2 14B training in clean bf16 on hardware that otherwise can't run it at all**, or only with the heaviest of the `special/low_vram_training/` tricks (gradient_checkpointing_offload + offload_models rotation + dataset_repeat caching across two stages). Most users have 24 GB cards, not 32 GB ones.

We don't have a 2× 24 GB rig on the bench right now to post measured numbers, but the architecture (~7 GB DiT half on each + bf16 T5 on cuda:0) leaves comfortable headroom on 24 GB. Happy to validate on 2× 3090/4090 if a maintainer has access and wants concrete numbers before merge.
